README.md

# XoopsSecure
[![Software License](https://img.shields.io/badge/license-GPL-brightgreen.svg?style=flat)](LICENSE) 

Xoops Secure is a module which from Admin side helps you to keep your installations clear from malware, virus and helps you to check for other security related items.




## Authors

- [@djculex](https://www.github.com/djculex)






## Features

- Scan / create index files
- Check / set defined files and folders for correct filer write permissions
- Scan for malware issues
- In development check Code Standards (PSR-12 & Xoops Standards)
- Check server for PHP/Mysql/Apache config issues
- Check for any reported Xoops security issues in CVR
- Set automatic scan for modified files and email admin report
- Keep an eye on configured GitHub repositories for latest releases
- Automatically / Manually create backup of files and MySql tables




## Minimum Requirements

- Xoops 2.5.10
- PHP 8.0
- MySql > 7.0




## Installation

Install as a normal Xoops Module

```PHP
  Admin -> install
  Admin -> xoopsSecure
```
    


docs/changelog.txt
1.3.5 RC4 [2023/02/21 14:28:00]
==============================================================
- Doc_comment vars defined by header include
- Rename classes inside namespaces
- Rename class files
- Adjust CS scan to disregard function/var naming.
- Fix temp folder + file not deleted in back up.

1.3.4 RC3 [2023/02/02 18:10:00]
==============================================================
- FIX : Multible fixes incl. (undefined constants, types, undefined class etc.) thanks Mamba
- FIX : Min version php 8.0 & missing language constant (thanks GregMage)

1.3.3 RC3 [2023/01/21 14:30:00]
==============================================================
- FIX : Using $_SERVER['server_protocol'] to create index files. Hardcoded http
  version seem to lock files sometimes giving 403 error when reading file content.
- ADD : Error logging to get server errors logged while scanning.


1.3.2 RC2 [2023/01/17 17:23:00]
==============================================================
- FIX : Backup paths as not all sub folders was included
- ADD : a function to call automatic backup if logged in as Admin
- Fix : Should we run cron and backup always was true
- Fix : timestamp for last cron scan was a string while int was expected.
- FIX : Folder for backup had wrong permissions.
- FIX : SpamScanner Check if folders are readable.
- FIX : After manual backup open in browser.
 


1.3.1 RC2 [2023/01/15 18:23:00]
==============================================================
- Additional cleaning up, renewing, moving and updating.. and
- ADD : agent to respond to ajax
- ADD : check for automatic scans
- ADD : Special file for Coding standard checks
- ADD : Bootstrap and bootbox functions and styles
- ADD : Logging for scan issues and scans dates
- ADD : Using GitHub api for repository updates
- ADD : Benchmarking class for later use
- Separating language files by class and added some new
- Including recursively to allow keeping class files in folders
- new and updated templates


1.1.0 [2022/11/04 12:31:31]
==============================================================
 - Major update by (Culex)


1.0.3 [2016/01/18 00:00:00]
==============================================================
 - Adjustments by (Mamba)


1.0.1 [2015/06/15 00:00:00]
==============================================================
 - Adjustments by (Mamba)


1.0.0 [2015/03/25 00:00:00]
==============================================================
 - Original release xoopsSecure (Culex)
 - Rebirth of the index scan idea but started over
 
 Adoption and namechange from IndexFile
 =============================================================
 - first release 2009/07/09 - last updated 2010/01/18
 - Credits Mamba, Voltan, Kris_fr, CJ, Reintjan, DonCurioso,
    Burning & Ghia just not to forget :-)

admin/scan.php
<?php
/**
 * ****************************************************************************
 * marquee - MODULE FOR XOOPS
 * Copyright (c) Hervé Thouzard (http://www.herve-thouzard.com)
 *
 * You may not change or alter any portion of this comment or credits
 * of supporting developers from this source code or any supporting source code
 * which is considered copyrighted (c) material of the original comment or credit authors.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * @copyright       Hervé Thouzard (http://www.herve-thouzard.com)
 * @license         http://www.fsf.org/copyleft/gpl.html GNU public license
 * @package         marquee
 * @author 			Hervé Thouzard (http://www.herve-thouzard.com)
 *
 * Version : $Id:
 * ****************************************************************************
 */


require_once dirname(dirname(dirname(dirname(__FILE__)))) . '/include/cp_header.php';
include_once dirname(__FILE__) . '/admin_header.php';
include_once (XOOPS_ROOT_PATH.'/class/template.php');
require_once XOOPS_ROOT_PATH . '/modules/xoopssecure/class/log.php';
require_once XOOPS_ROOT_PATH . '/modules/xoopssecure/class/scan.php';
require_once XOOPS_ROOT_PATH . '/modules/xoopssecure/class/mech.php';
include_once XOOPS_ROOT_PATH . '/modules/xoopssecure/class/backup.php';


$scan = new xoopsSecure_scan;
$log  = new xoopsSecure_log;
$sys = new xoopsSecure_mech;
$zip = new xoopsSecureZipper;

global $xoopsUser, $xoopsLogger;
$xoopsLogger->activated = false;
//error_reporting(E_ALL); 

set_time_limit(999999);
$task = intval($_GET['scantype']);
$scanref = intval($_GET['ref']);
$scanurl = $_GET['scanurl'];
$scan->inittime = $_GET['scanstart'];

// Quick scan = 1
if ($task == 1) {
    $scan->scantype = $scanref;
    $info = $scan->getnewfileinfo ($url = $scanurl);
    header('Content-type: application/json');
    echo "{\"result\":$info}";  
}

// Developer scan
if ($task == 2) {
    $type = xoopssecure_GetModuleOption('checkstylestyle', $repmodule='xoopssecure');
    $_SERVER['argc'] = 5;
    
    $standard = '--standard='.$type[0];
    $report = "--report=full";
    $reportWidth = "--report-width=110";
    
    $url = $_GET['path'];

    $_SERVER['argv'] = array("phpcs.php",$standard,$url,$report,$reportWidth);
    echo "<div class='xoopssecure_csf_report_cont'><p>";
    echo sprintf( _AM_XOOPSSECURE_DEVELOPERSCANRESULTTITLE, $url, $type[0] );
    echo "</p>";
    echo "<div id = 'xoopssecure_showoksign' style='display:none;'><img src = '../assets/images/okstamp.PNG' "
        ." height='100px' width= '100px' /></div>";
    echo '<div class="xoopssecure_csf_report"><code>';
    include XOOPS_ROOT_PATH. '/modules/xoopsSecure/plugins/codesniffer/CodeSniffer/phpcs.php';
    echo '</code></div>';
    echo '</div>';
}


// full scan = 3
if ($task == 3) {
    $scan->scantype = $scanref;
    if ($scanref == 3) {
        $scan->emptyFiles ();
    }
    if ($scanurl !== XOOPS_ROOT_PATH) {
        $scan->getnewfileinfo ($url = XOOPS_ROOT_PATH);
    }
    $info = $scan->getnewfileinfo ($url = $scanurl);
    header('Content-type: application/json');
    echo "{\"result\":$info}";    
}

// Single file scan = 4
if ($task == 4) {
    $scan->scantype = $scanref;
    $file = $_GET['file'];
    //$scan->savefileinfo ($file);
    $scanref == 3 ? $scan->getLines (json_decode($file), $scanref) : $scan->gethashes(json_decode($file), $scanref);
}

// indexfile creating
if ($task == 5) {
    $dir = $_GET['file'];
    // auto indexfile creation
    $scan->autoindexcreate == 1 ? $scan->createindexfiles($dir) : '';
}

// Chmodding
if ($task == 6) {
    $dir = $_GET['file'];
    // auto indexfile creation
    header('content-type: application/json; charset=utf-8');
    if ($scan->autochmod[0] == 1) {
        echo $scan->recursiveChmod (str_replace ("\\", "/", $dir), $scanref);
    } else {
        break;
    }
}

// create .httaccess
if ($task == 7) {
    $dir = $_GET['file'];
    // auto indexfile creation
    $scan->autoHtaccessCreate == 1 ? $scan->createHttaccess($dir) : '';
}

// Delete from log
if ($task == 8) {
    header('content-type: application/json; charset=utf-8');
    $id = intval($_GET['id']);
    $table = ($_GET['table']);
    $scan->deleteById($id, $table);
}

// add to ignore
if ($task == 9) {
	$file = $_GET['file'];
	$linenumber = $_GET['ln'];
	$scan->Ignore($file, $linenumber);
	header('content-type: application/json; charset=utf-8');
}

// Empty ignore
if ($task == 10) {
	header('content-type: application/json; charset=utf-8');
}

// indexfile check
if ($task == 11) {
    $dir = $_GET['file'];
    // check
    $scan->checkMissingIndexfile($dir);
}

// Set develop path
if ($task == 12) {
    //$dir = $_GET['developpath'];
    $scan->developPath = $_GET['developpath'];
    // check
    //$scan->checkMissingIndexfile($dir);
}

// Set Full scan path
if ($task == "12f") {
    //$dir = $_GET['developpath'];
    $scan->urlToScan = $_GET['urlToScan'];
    // check
    //$scan->checkMissingIndexfile($dir);
}

// Show Malware log
if ($task == 13) {
    $log->getIssues ('file');
    $log->getIssues ('dir');
    $xoopsTpl->display(XOOPS_ROOT_PATH .'/modules/xoopssecure/templates/admin/showscan.tpl'); 
}

// ADD file / dir to ignore
if ($task == 14) {
    $file = $_GET['file'];
    $type = $_GET['ref'];
    $val = $_GET['val'];
    $table = isset($_GET['table']) ? $_GET['table'] : '';
    if ($val != 'clear') {
        $scan->ignoreFile ($file, $type, $val);
    } else {
        $log->clear ($file);
    }
}

// ADD file / dir to chmod reset list
if ($task == 16) {
    $file = $_GET['file'];
    $type = $_GET['ref'];
    $val = $_GET['val'];
    $table = $_GET['table'];    
    $scan->ignoreFile ($file, $type, $val);
}

// Get list back of ignore dirs / files
if ($task == 17) {
    $genus = $_GET['ref'];
    $val = $_GET['val'];
    $scan->getIgnores ($genus, $val);
}

// do backup
if ($task == 18) {
	if ($this->backuptype != 'none') {
		$zip->doZip ($zip->archive, $zip->dirToBackup);
		$nl = str_replace(XOOPS_ROOT_PATH,XOOPS_URL,$zip->archive);
		$link  = "<a id = 'xoopssecure_bdl' href = '".$nl."'> "
		. _AM_XOOPSSECURE_BACKUP_LINK_DESC." "
		. $nl
		. "</a>";
		echo $link;
	}
}

// download the zip file
if ($task == 19) {
	if ($scan->autobackup = 1 OR $scanref = 18) {
		xoopssecure_DownloadFile($zip->archive);
	}
}

//delete backupfolder from server
if ($task == 20) {
	if ($scan->autobackup = 1 OR $scanref = 18) {
		//xoopssecure_deleteBackupFolder ($zip->archive);
		xoopssecure_deleteFolder ($zip->dest);
	}
}